src/domain/core/src/services/export_service.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::Path;
+
+use datafusion::dataframe::DataFrame;
+use internal_error::{BoxedError, InternalError};
+use thiserror::Error;
+
+use crate::*;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Service
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait ExportService: Send + Sync {
+    async fn export_to_fs(
+        &self,
+        df: DataFrame,
+        path: &Path,
+        format: &ExportFormat,
+        records_per_file: Option<usize>,
+    ) -> Result<u64, ExportError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Clone, strum::Display, Debug, strum::EnumIter)]
+pub enum ExportFormat {
+    #[strum(to_string = "parquet")]
+    Parquet,
+
+    #[strum(to_string = "csv")]
+    Csv,
+
+    #[strum(to_string = "ndjson")]
+    NdJson,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Errors
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Error)]
+#[error("Invalid SQL query: {context}")]
+pub struct ExportQueryError {
+    pub context: String,
+    #[source]
+    pub source: BoxedError,
+}
+
+#[derive(Debug, Error)]
+pub enum ExportError {
+    #[error(transparent)]
+    Internal(
+        #[from]
+        #[backtrace]
+        InternalError,
+    ),
+    #[error(transparent)]
+    DataFusionError(DataFusionError),
```

Comment:
Aren't all DF errors still internal errors? Why do we need to differentiate them?

Reply:
`ExportError::Internal` is thrown when we failed to calculate exported rows.
For all other cases we throw `DataFusionError`. I took QueryService as an example, so new command and changes to `sql` one conforms existent interface. 
With this approach we sometimes may show user not useful details. But if we want to improve this, we need to be careful with filtering DataFusion errors in order to provide users with details when their sql is invalid, or they cannot store data to a specific location. I think we need to make changes to both Query and External service at the same time. And if we'll go with this direction, I guess it makes sense to implement it in scope of separate ticket.

Reply:
For the context, here're some of DataFusion errors exposed to users.
Triggered from QueryService:
```bash
❯ kamu sql --command "select random_field_name from 'test.data2'" --output-format parquet --output-path ./out/parquet/1
Error:
  0: DataFusion error
  1: Schema error
  2: No field named random_field_name. Valid fields are "test.data2".offset, "test.data2".op, "test.data2".system_time, "test.data2".event_time, "test.data2".uuid1.
  
❯ kamu sql --command "select * from 'test.data2' qwe 212" --output-format parquet --output-path /nix/1.parquet
Error:
  0: DataFusion error
  1: SQL error: ParserError("Expected end of statement, found: 212")
  2: sql parser error: Expected end of statement, found: 212

❯ kamu sql --command "select * from 'does.not.exists'" --output-format parquet --output-path ./out/parquet/1
Error:
  0: DataFusion error
  1: Error during planning: table 'kamu.kamu.does.not.exists' not found
```
Triggered from ExportService:
```bash❯ kamu sql --command "select 1/0" --output-format parquet --output-path ./out/parquet/1
Error:
  0: DataFusion error
  1: Arrow error
  2: Divide by zero error
  
❯ kamu sql --command "select * from 'test.data2'" --output-format parquet --output-path /out/parquet/1
Error:
  0: DataFusion error
  1: IO error
  2: Generic LocalFileSystem error
  3: Unable to create dir /out/parquet/1
  4: Read-only file system (os error 30)
  
❯ kamu sql --command "select * from 'test.data2'" --output-format parquet --output-path /nix/1.parquet
Error:
  0: DataFusion error
  1: IO error
  2: Generic LocalFileSystem error
  3: Unable to open file /nix/1.parquet#1
  4: Permission denied (os error 13)  
  
❯ kamu sql --command "select * from 'test.data2' limit -1" --output-format parquet --output-path /nix/1.parquet
Error:
  0: DataFusion error
  1: Optimizer rule 'eliminate_limit' failed
caused by
Error during planning: LIMIT must be >= 0, '-1' was provided
  2: Error during planning: LIMIT must be >= 0, '-1' was provided
```

Reply:
All errors you provided as examples above look like planning-phase errors to me, so If I'm not mistaken they should all be raised from the `QueryServiceImpl` while we are setting up the logical plan, not from `ExportServiceImpl`.

Here it makes sense for `QueryService` to return `DataFusionError` instead of `InternalError` because very likely all errors will be caused by the invalid user input.

This logical plan in form of a `DataFrame` is then passed into the `ExportService` where we add a few extra steps:
- repartition + file sync - a planning error on at phase will indicate we badly messed up the code
- finally we execute the plan - an error at this phase will likely be IO-related (out of space, permissions etc).

In both cases there isn't much what user can do about them. And the latter execution-phase errors when used in `kamu-node` context can reveal details of internal infrastructure (e.g. some S3 bucket names or object keys).

Reply:
Some errors like division by zero and invalid limit value are spotted on execution phase (in ExportService). I believe there are a few more useful ones I didn't spot. And also IO errors in example above looks useful as well. I'd preserve them either as DataFusion errors, or create specific ones, and would map corresponding DataFusion errors to that new ones.